### PR TITLE
feat(backend): Add ExecutionClient to access ExecutionSpec

### PR DESCRIPTION
**Pull request by yhwang**
**Description of your changes:**
issue: #6888

Add ExecutionClient, ExecutionInterface to abstract the client
functions for ExecutionSpec. Replace WorkflowClient/WorkflowInterface
in all backend components to use ExecutionClient/ExecutionInterface.
Only add needed functions into ExecutionClient/ExecutionInterface.

Signed-off-by: Yihong Wang <yh.wang@ibm.com>

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

**Comment by yhwang** [comment]
This should be the last PR to provide abstract APIs for orchestration engines (currently Argo). In this PR, two majors interfaces are created to access the runtime engine's CRs:
- ExecutionInterface: provides APIs to access runtime engine's CRs, including `CRUD`, `List`, and `Patch`
- ExecutionInformer: similar to SharedInformer, provides event registration, `List`, and `Get` APIs

Then refactor backend components to use these two interfaces and remove the direct dependency of Argo go pkg.

**Comment by yhwang** [comment]
@chensun please review this PR, thanks.

**Comment by yhwang** [comment]
Hi @chensun , I did a rebase yesterday. Can you review this PR? Thanks.

File: backend/src/crd/pkg/apis/scheduledworkflow/v1beta1/zz_generated.deepcopy.go [hunk 1]
```diff
@@ -323,7 +323,6 @@ func (in *WorkflowResource) DeepCopyInto(out *WorkflowResource) {
 		*out = make([]Parameter, len(*in))
 		copy(*out, *in)
 	}
-	in.Spec.DeepCopyInto(&out.Spec)
```
Comment (chensun) [inline backend/src/crd/pkg/apis/scheduledworkflow/v1beta1/zz_generated.deepcopy.go:326]: For my education how is this file updated?

**Review by chensun — COMMENTED** [review]
/lgtm

/cc @Linchin to take a look as well

File: backend/src/crd/pkg/apis/scheduledworkflow/v1beta1/zz_generated.deepcopy.go (hunk 1 above)
Comment (yhwang) [reply backend/src/crd/pkg/apis/scheduledworkflow/v1beta1/zz_generated.deepcopy.go:326]: the spec is no longer Argo's Workflow.Spec. In order to remove the dependency, it's an JSON string representing the orchestration engine's CR now. so no DeepCopy for a string

**Comment by yhwang** [comment]
@chensun and @Linchin resolved the conflicts and rebased again.

**Review by chensun — COMMENTED** [review]
/lgtm
/approve

**Comment by yhwang** [comment]
@chensun thanks for the review/comments

**Comment by Linchin** [comment]
Hi @yhwang, it looks like this change is causing some upgrade issues, which lead to status of newly created runs not shown. Would you check it out and see if you could fix the issue? If you wouldn't be able to do this recently, we might have to revert the change of this PR. Here are the steps to reproduce the issue:

* Deploy 2.0.0-alpha.3
* Create recurring run using v1 
* Create recurring run using v2
* Enable both recurring run for 10 min 
* Upgrade to 2.0.0-alpha.4.
* Create a v1 run, no run status, and the pipeline graph does not load.
* Create a v2 run, also no run status. But you can see the pipeline graph.
* But if I deleted the recurring runs, (and possibly need to create a new run) the greyed out run status become green.

Feel free to ping me if you have any question.

**Comment by yhwang** [comment]
@Linchin Thanks for the info, let me recreate the issue you mentioned and start to debug

**Pull request merged**